daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala

```diff
@@ -174,61 +183,146 @@ private[lf] final class ValueTranslator(
               // records
-              case ValueRecord(mbId, flds) =>
-                mbId.foreach(id =>
-                  if (id != tyCon)
-                    typeError(
-                      s"Mismatching record id, the type tells us $tyCon, but the value tells us $id"
-                    )
-                )
+              case ValueRecord(mbId, sourceElements) =>
+                checkUserTypeId(tyCon, mbId)
                 val lookupResult = handleLookup(pkgInterface.lookupDataRecord(tyCon))
-                val recordFlds = lookupResult.dataRecord.fields
-                // note that we check the number of fields _before_ checking if we can do
-                // field reordering by looking at the labels. this means that it's forbidden to
-                // repeat keys even if we provide all the labels, which might be surprising
-                // since in JavaScript / Scala / most languages (but _not_ JSON, interestingly)
-                // it's ok to do `{"a": 1, "a": 2}`, where the second occurrence would just win.
-                if (recordFlds.length != flds.length) {
-                  typeError(
-                    s"Expecting ${recordFlds.length} field for record $tyCon, but got ${flds.length}"
-                  )
-                }
+                val targetFieldsAndTypes = lookupResult.dataRecord.fields
                 val subst = lookupResult.subst(tyArgs)
-                val fields = labeledRecordToMap(flds) match {
-                  case None =>
-                    (recordFlds zip flds).map { case ((lbl, typ), (mbLbl, v)) =>
-                      mbLbl.foreach(lbl_ =>
-                        if (lbl_ != lbl)
-                          typeError(
-                            s"Mismatching record field label '$lbl_' (expecting '$lbl') for record $tyCon"
-                          )
-                      )
-                      val replacedTyp = AstUtil.substitute(typ, subst)
-                      lbl -> go(replacedTyp, v, newNesting)
+
+                if (config.enableUpgrade) {
+                  // We handle the (non-)upgrade cases for record separately
+                  // In case of upgrade, we require the field are record.
+
+                  // This code implements the compatibility transformation used for up/down-grading
+                  // And handles the cases:
+                  // - UPGRADE:   numT > numS : creates a None for each missing fields.
+                  // - DOWNGRADE: numS > numT : drops each extra field, ensuring it is None.
+                  //
+                  // When numS == numT, we wont hit the code marked either as UPGRADE or DOWNGRADE,
+                  // although it is still possible that the source and target types are different,
+                  // but since we don't consult the source type (may be unavailable), we wont know.
+
+                  val numS: Int = sourceElements.length
+                  val numT: Int = targetFieldsAndTypes.length
+
```

Comment:
Do we have plans to merge this code back with the code in `Speedy.scala` which it was cut-and-pasted from?

Reply:
I have no concrete plan. If you want to factorize this code with the importValue, we probably have to try to factorize the whole function. I am not sure how to do that.
I suggest we revisit the question  when we have more advanced in the implementation of upgrade. 
Let's me add a TODO.

Reply:
There is currently no concrete plan. 
Let's revisit the question when we are more advance with the upgrade implementation. 
I add a TODO.